### DISCUSSION TITLE
Add support for Hi-res

### DISCUSSION THREAD

**Discussion by Maverick565** [GioF71/tidal-connect#147] [Ideas]
Hi!

Did you try to add this variable param to configure.sh to get hi-res? I saw this on volumio.

`soundQuality=HI_RES`

**Comment by GioF71** [comment]
Hello, unfortunately this specific binary does not have this option.
I have run the help switch and there is nothing obvious that I can enable in order to enable hires.
It's not the same as the software (or service) used by Volumio.
If you want/need hi-res, I can only suggest you to consider the options that are listed in the README.md.

**Reply by Maverick565** [reply]
Is there a possible to control hardware device volume instead of using software volume?

**Reply by GioF71** [reply]
Well not that I know of... I found that I could add software volume via alsa configuration, but I will check if something exists for hardware volume.

Here are the available switches:

```
usage:
  tidal_connect_application  options

where options are:
  --log-level, -l <log_level_number>        Log level
                                                  0: OFF
                                                  1: CRITICAL
                                                  2: ERROR
                                                  3: WARNING
                                                  4: INFO
                                                  5: DEBUG
                                                  6: TRACE
  --netif-for-deviceid                      Network interface for device
  <network_interface_name_string>
  --additional-seed <string>                Additional seed value for device
                                            ID
  --codec-mpegh <true|false>                Codec MPEGH capability flag for
                                            mdns
  --codec-mqa <true|false>                  Codec MQA capability flag for
                                            mdns
  --codec-atmos <true|false>                Codec Dolby Atmos capability flag
                                            for mdns
  --model-name, -m <model_name_string>      Model name for mdns
  --friendly-name, -f                       Friendly name for mdns
  <friendly_name_string>
  --websocket-port, -p <port_number>        Websocket port
  --disable-app-security <true|false>       Disable app security
  --enable-websocket-log                    Websocket log level
  <websocket_log_level_number>                    0: Disable
                                                  1: AccessCore
                                                  2: AccessCore_Controll
                                                  3: All
  --disable-web-security <true|false>       Disable web security
  --controller-unix-socket-path             Speaker communicator Unix socket
  <unix_socket_path>                        path
  --web-ca-path <web_ca_cert_path>          Web certificate path
  --clientid <tidal_client_id_string>       Client id
  --tc-certificate-path                     tidal connect certificate path
  <tidal_certificate_string>
  --enable-debug-port <true|false>          Enable debug port
  --playback-device                         Name of device to use for
  <playback_device_string>                  playback
  --enable-mqa-passthrough <true|false>     Enable MQA passthrough
  --stream-fetch-limit <number_of_bytes>    Set the limit in bytes for how
                                            much of a track to fetch
  --audio-output-buffer                     Set the buffer size to be used
  <number_of_samples>                       for decoded samples
  -?, -h, --help                            display usage information

```

**Reply by GioF71** [reply]
I just tried, I can set hardware volume to control the volume of Tidal Connect, but of course I did that using alsamixer from the command line. The Tidal apps AFAIK only access the control named "Master" and that's not how the hardware volume controls are called. Example for my Topping D10:

```
osmc@KodiLivingD10:~/git/tidal-connect$ cat /proc/asound/cards
 0 [DAC            ]: USB-Audio - YULONG Audio 384k DSD DAC
                      YULONG Audio YULONG Audio 384k DSD DAC at usb-0000:01:00.0-1.4, high speed
 1 [D10            ]: USB-Audio - D10
                      Topping D10 at usb-0000:01:00.0-1.3, high speed
 2 [vc4hdmi0       ]: vc4-hdmi - vc4-hdmi-0
                      vc4-hdmi-0
 3 [vc4hdmi1       ]: vc4-hdmi - vc4-hdmi-1
                      vc4-hdmi-1
osmc@KodiLivingD10:~/git/tidal-connect$ amixer -c 1 controls
numid=4,iface=CARD,name='Topping Internal Clock Validity'
numid=2,iface=MIXER,name='D10  Playback Switch'
numid=3,iface=MIXER,name='D10  Playback Volume'
numid=1,iface=PCM,name='Playback Channel Map'
```

I don't know if there is a way to convince this app to use a different mixer. From the available settings I see, I would say no, but I would be glad to be corrected

**Reply by Maverick565** [reply]
Thats how it looks in my Sound Blaster card. 

![Screenshot_20240306_171228_Termius.jpg](https://github.com/GioF71/tidal-connect/assets/54276954/378cdcff-0cda-496c-a67a-31f9e4d6c0c9)

**Reply by Maverick565** [reply]
![Screenshot_20240306_171430_Termius.jpg](https://github.com/GioF71/tidal-connect/assets/54276954/fb20809a-ea45-46f0-ba90-52329461adeb)

**Reply by CarelessEvening** [reply]
> Hello, unfortunately this specific binary does not have this option. I have run the help switch and there is nothing obvious that I can enable in order to enable hires. It's not the same as the software (or service) used by Volumio. If you want/need hi-res, I can only suggest you to consider the options that are listed in the README.md.

Heya, taking this space to thank you for this solution! Super happy to see and be able to play to my diy streamer straight from the Tidal app using connect. I'd like to ask for a clarification about the alternatives specified in the readme - do any of them allow for Tidal Connect + HiRes? I saw you metioned HiRes is available i some of them, but wasn't sure about the 'Tidal Connect' availability. Thanks again for your help!

**Reply by GioF71** [reply]
Hello @CarelessEvening and thank you for the kind words.
Unfortunately, the only real alternative that can be installed on a DIY device is Volumio, which means that you need to pay an extra subscription in order to have Tidal Connect. I have tried it and I must say that it works well.
Many of the other alternatives support HiRes, but not Tidal Connect, so if this is a mandatory requirement, the options are quite limited.

If you want Tidal hires on your DIY streamer, my top picks are:

- BubbleUPnP (worth paying for the app IMO)? The Tidal integration is really nice.
- Tidal plugin for upmpdcli, which is free (provided you have a running Tidal subscription of course) and currently works in hires. Disclaimer, I am the author of that plugin.
- Music Assistant: the Tidal plugin works in hi-res, but I'd suggest to use Squeezelite as the player for better results.
- Mopidy-Tidal also supports Tidal-Hires, needs to run with in the playing device though. Not a showstopper but it might be a bit heavy in case you have a Pi3. I have a docker container for this solution as mentioned in the alternatives.

**Comment by GioF71** [comment]
Ok for the mixer names, try:

for S4:

`amixer -c 2 controls`

For K5

`amixer -c 3 controls`

**Reply by Maverick565** [reply]
![Screenshot_20240306_182915_Termius.jpg](https://github.com/GioF71/tidal-connect/assets/54276954/c46ae6e2-b4b8-491f-b47d-bede60ba5255)

S4.

**Comment by GioF71** [comment]
Yes, infact no control is named Master
